File: eznarratives_reflex_kivy/tests/test_e2e.py
"""
End-to-End Tests for EZ Narratives
=================================

This module contains end-to-end tests for the EZ Narratives application.
These tests use Puppeteer to automate browser interactions and verify
the application's functionality from a user's perspective.
"""

import os
import sys
import pytest
import asyncio
from dotenv import load_dotenv
from pyppeteer import launch

# Add the parent directory to the path so we can import our modules
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Load environment variables
load_dotenv(".env.local")


@pytest.mark.asyncio
async def test_login_flow():
    """Test the login flow."""
    # Launch the browser
    browser = await launch(headless=True)
    page = await browser.newPage()
    
    try:
        # Navigate to the login page
        await page.goto('http://localhost:3004/login')
        
        # Wait for the page to load
        await page.waitForSelector('input[id="email"]')
        
        # Fill in the login form
        await page.type('input[id="email"]', 'test@example.com')
        await page.type('input[id="password"]', 'password123')
        
        # Click the login button
        login_button = await page.querySelector('button[type="submit"]')
        await login_button.click()
        
        # Wait for navigation to dashboard
        await page.waitForNavigation()
        
        # Verify that we're on the dashboard page
        current_url = page.url
        assert '/dashboard' in current_url, f"Expected to be on dashboard page, but got {current_url}"
        
    finally:
        # Close the browser
        await browser.close()


@pytest.mark.asyncio
async def test_ems_narrative_generation():
    """Test generating an EMS narrative."""
    # Launch the browser
    browser = await launch(headless=True)
    page = await browser.newPage()
    
    try:
        # Log in first
        await page.goto('http://localhost:3004/login')
        await page.waitForSelector('input[id="email"]')
        await page.type('input[id="email"]', 'test@example.com')
        await page.type('input[id="password"]', 'password123')
        login_button = await page.querySelector('button[type="submit"]')
        await login_button.click()
        await page.waitForNavigation()
        
        # Navigate to the EMS tab
        ems_tab = await page.querySelector('button:has-text("EMS")')
        await ems_tab.click()
        
        # Wait for the EMS form to load
        await page.waitForSelector('input[id="unit"]')
        
        # Fill in the EMS form
        await page.type('input[id="unit"]', 'Medic 1')
        await page.type('input[id="dispatch_reason"]', '123 Main St for chest pain')
        await page.select('select[id="patient_sex"]', 'Male')
        await page.type('input[id="patient_age"]', '65')
        await page.type('input[id="chief_complaint"]', 'Chest pain')
        
        # Click the generate narrative button
        generate_button = await page.querySelector('button:has-text("Generate Narrative")')
        await generate_button.click()
        
        # Wait for the narrative to be generated
        await page.waitForFunction(
            'document.querySelector("div.narrative-text").textContent.includes("EMS NARRATIVE")',
            {'timeout': 10000}
        )
        
        # Verify that a narrative was generated
        narrative_text = await page.evaluate('document.querySelector("div.narrative-text").textContent')
        assert "EMS NARRATIVE" in narrative_text, "Expected EMS narrative to be generated"
        
    finally:
        # Close the browser
        await browser.close()


@pytest.mark.asyncio
async def test_fire_narrative_generation():
    """Test generating a Fire narrative."""
    # Launch the browser
    browser = await launch(headless=True)
    page = await browser.newPage()
    
    try:
        # Log in first
        await page.goto('http://localhost:3004/login')
        await page.waitForSelector('input[id="email"]')
        await page.type('input[id="email"]', 'test@example.com')
        await page.type('input[id="password"]', 'password123')
        login_button = await page.querySelector('button[type="submit"]')
        await login_button.click()
        await page.waitForNavigation()
        
        # Navigate to the Fire tab
        fire_tab = await page.querySelector('button:has-text("Fire")')
        await fire_tab.click()
        
        # Wait for the Fire form to load
        await page.waitForSelector('input[id="unit"]')
        
        # Fill in the Fire form
        await page.type('input[id="unit"]', 'Engine 3')
        await page.select('select[id="emergency_type"]', 'Structure Fire')
        await page.type('textarea[id="additional_info"]', 'Two-story residential structure with heavy smoke showing from second floor.')
        
        # Click the generate report button
        generate_button = await page.querySelector('button:has-text("Generate NFIRS Report")')
        await generate_button.click()
        
        # Wait for the report to be generated
        await page.waitForFunction(
            'document.querySelector("div.narrative-text").textContent.includes("FIRE INCIDENT REPORT")',
            {'timeout': 10000}
        )
        
        # Verify that a report was generated
        report_text = await page.evaluate('document.querySelector("div.narrative-text").textContent')
        assert "FIRE INCIDENT REPORT" in report_text, "Expected Fire report to be generated"
        
    finally:
        # Close the browser
        await browser.close()


@pytest.mark.asyncio
async def test_chat_functionality():
    """Test the chat functionality."""
    # Launch the browser
    browser = await launch(headless=True)
    page = await browser.newPage()
    
    try:
        # Log in first
        await page.goto('http://localhost:3004/login')
        await page.waitForSelector('input[id="email"]')
        await page.type('input[id="email"]', 'test@example.com')
        await page.type('input[id="password"]', 'password123')
        login_button = await page.querySelector('button[type="submit"]')
        await login_button.click()
        await page.waitForNavigation()
        
        # Navigate to the Chat tab (should be default)
        chat_tab = await page.querySelector('button:has-text("Chat")')
        await chat_tab.click()
        
        # Wait for the chat input to load
        await page.waitForSelector('textarea[placeholder*="Type your message"]')
        
        # Send a message
        await page.type('textarea[placeholder*="Type your message"]', 'Hello, can you help me with an EMS narrative?')
        send_button = await page.querySelector('button[aria-label="Send message"]')
        await send_button.click()
        
        # Wait for a response
        await page.waitForFunction(
            'document.querySelectorAll("div.message").length > 1',
            {'timeout': 10000}
        )
        
        # Verify that a response was received
        messages = await page.evaluate('Array.from(document.querySelectorAll("div.message")).map(m => m.textContent)')
        assert len(messages) > 1, "Expected at least one response message"
        
    finally:
        # Close the browser
        await browser.close()


if __name__ == "__main__":
    # Run the tests
    pytest.main(["-xvs", __file__])

File: eznarratives_reflex_kivy/tests/test_e2e.js
/**
 * End-to-End Test for EZ Narratives
 * 
 * This script uses Puppeteer to test the web application.
 * It performs the following steps:
 * 1. Opens the Reflex web server
 * 2. Logs in
 * 3. Fills in the EMS form
 * 4. Generates a narrative
 * 5. Switches to the Fire tab
 * 6. Fills in the Fire form
 * 7. Generates a fire narrative
 */

const puppeteer = require('puppeteer');
require('dotenv').config({ path: '.env.local' });

// Test configuration
const config = {
  baseUrl: 'http://localhost:3000',
  credentials: {
    email: process.env.TEST_USER_EMAIL || 'test@example.com',
    password: process.env.TEST_USER_PASSWORD || 'password123'
  },
  timeout: 30000 // 30 seconds
};

// Helper function to wait for a specific time
const wait = (ms) => new Promise(resolve => setTimeout(resolve, ms));

(async () => {
  console.log('Starting E2E test...');
  
  // Launch the browser
  const browser = await puppeteer.launch({
    headless: false, // Set to true for CI environments
    args: ['--no-sandbox', '--disable-setuid-sandbox'],
    defaultViewport: { width: 1280, height: 800 }
  });
  
  const page = await browser.newPage();
  
  try {
    // Step 1: Navigate to the application
    console.log('Navigating to the application...');
    await page.goto(config.baseUrl, { waitUntil: 'networkidle2', timeout: config.timeout });
    
    // Step 2: Log in
    console.log('Logging in...');
    await page.waitForSelector('input[type="email"]');
    await page.type('input[type="email"]', config.credentials.email);
    await page.type('input[type="password"]', config.credentials.password);
    await page.click('button[type="submit"]');
    
    // Wait for dashboard to load
    await page.waitForSelector('[data-testid="dashboard"]', { timeout: config.timeout });
    console.log('Successfully logged in!');
    
    // Step 3: Fill in EMS form
    console.log('Filling in EMS form...');
    // Click on EMS tab
    await page.click('[data-testid="ems-tab"]');
    await wait(1000);
    
    // Fill in dispatch section
    await page.type('[data-testid="unit-input"]', 'Medic 1');
    await page.type('[data-testid="dispatch-reason-input"]', '123 Main St for chest pain');
    
    // Fill in patient information
    await page.select('[data-testid="patient-sex-select"]', 'Male');
    await page.type('[data-testid="patient-age-input"]', '65');
    await page.type('[data-testid="chief-complaint-input"]', 'Chest pain');
    await page.type('[data-testid="duration-input"]', '30 minutes');
    await page.type('[data-testid="patient-presentation-input"]', 'Patient found sitting upright in chair, clutching chest.');
    
    // Fill in assessment section
    await page.click('[data-testid="next-section-button"]');
    await wait(500);
    
    // Fill in treatment section
    await page.click('[data-testid="next-section-button"]');
    await wait(500);
    await page.type('[data-testid="treatment-provided-input"]', 'Administered 324mg aspirin PO, established IV access.');
    
    // Fill in transport section
    await page.click('[data-testid="next-section-button"]');
    await wait(500);
    await page.type('[data-testid="transport-destination-input"]', 'Memorial Hospital');
    
    // Step 4: Generate EMS narrative
    console.log('Generating EMS narrative...');
    await page.click('[data-testid="generate-narrative-button"]');
    
    // Wait for narrative to be generated
    await page.waitForSelector('[data-testid="narrative-text"]', { timeout: config.timeout });
    
    // Verify narrative was generated
    const emsNarrativeText = await page.$eval('[data-testid="narrative-text"]', el => el.textContent);
    console.log('EMS Narrative generated successfully!');
    console.log('Narrative preview:', emsNarrativeText.substring(0, 100) + '...');
    
    // Step 5: Switch to Fire tab
    console.log('Switching to Fire tab...');
    await page.click('[data-testid="fire-tab"]');
    await wait(1000);
    
    // Step 6: Fill in Fire form
    console.log('Filling in Fire form...');
    await page.type('[data-testid="unit-input"]', 'Engine 3');
    await page.select('[data-testid="emergency-type-select"]', 'Structure Fire');
    await page.type('[data-testid="additional-info-input"]', 'Two-story residential structure with heavy smoke showing from second floor.');
    
    // Step 7: Generate Fire narrative
    console.log('Generating Fire narrative...');
    await page.click('[data-testid="generate-fire-narrative-button"]');
    
    // Wait for narrative to be generated
    await page.waitForSelector('[data-testid="fire-narrative-text"]', { timeout: config.timeout });
    
    // Verify narrative was generated
    const fireNarrativeText = await page.$eval('[data-testid="fire-narrative-text"]', el => el.textContent);
    console.log('Fire Narrative generated successfully!');
    console.log('Narrative preview:', fireNarrativeText.substring(0, 100) + '...');
    
    console.log('E2E test completed successfully!');
  } catch (error) {
    console.error('E2E test failed:', error);
    
    // Take a screenshot of the failure
    await page.screenshot({ path: 'e2e-test-failure.png' });
    
    throw error;
  } finally {
    // Close the browser
    await browser.close();
  }
})();